File: video_ingestion.py

# Copyright (c) 2019 Intel Corporation.

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.


import threading
import queue
import time
import datetime
import json
import os
import signal
import logging
import argparse
from distutils.util import strtobool
import sys
from ingestor import Ingestor
from libs.base_filter import load_filter
from util.log import configure_logging, LOG_LEVELS
from libs.ConfigManager import ConfigManager
from util.msgbusutil import MsgBusUtil
from publisher import Publisher

# Default queue size for filter input queue for `no_filter` case
QUEUE_SIZE = 10


class VideoIngestion:

    def __init__(self, dev_mode, config_client):
        """Get the frames from camera or video, filters and add the results
        to the messagebus

        :param dev_mode: indicates whether it's dev or prod mode
        :type dev_mode: bool
        :param config_client: distributed store config client
        :type config_client: config client object
        """
        self.log = logging.getLogger(__name__)
        self.profiling = bool(strtobool(os.environ['PROFILING_MODE']))
        self.dev_mode = dev_mode
        self.app_name = os.environ["AppName"]
        self.config_client = config_client
        self._read_ingestor_filter_config()
        self.config_client.RegisterDirWatch("/{0}/".format(self.app_name),
                                            self._on_change_config_callback)

    def _print_config(self):
        self.log.info('ingestor_config: {}'.format(self.ingestor_config))
        if self.filter_config:
            self.log.info('filter name: {} filter config: {}'.format(
                self.filter_config["name"], self.filter_config))

    def _read_ingestor_filter_config(self):
        CONFIG_KEY_PATH = "/config"
        self.config = self.config_client.GetConfig("/{0}{1}".format(
            self.app_name, CONFIG_KEY_PATH))

        self.config = json.loads(self.config)
        self.ingestor_config = self.config["ingestor"]
        self.filter_config = self.config.get("filter", None)

        self.filter_input_queue = queue.Queue(
            maxsize=QUEUE_SIZE)

    def start(self):
        """Start Video Ingestion.
        """
        log_msg = "======={} {}======="
        self.log.info(log_msg.format("Starting", self.app_name))

        self._print_config()

        pub_topic = MsgBusUtil.get_topics_from_env("pub")

        if len(pub_topic) > 1:
            self.log.error("More than one publish topic in {} " +
                           "is not allowed".format(self.app_name))
            sys.exit(1)

        if self.filter_config:
            queue_size = self.filter_config["queue_size"]
            self.filter_input_queue = queue.Queue(
                maxsize=queue_size)
            self.filter_output_queue = queue.Queue(
                maxsize=queue_size)
        else:
            # for `no filter` cfg
            self.filter_output_queue = self.filter_input_queue

        pub_topic = pub_topic[0]
        self.publisher = Publisher(self.filter_output_queue, pub_topic,
                                   self.config_client, self.dev_mode)
        self.publisher.start()

        if self.filter_config:
            self.filter = load_filter(self.filter_config["name"],
                                      self.filter_config,
                                      self.filter_input_queue,
                                      self.filter_output_queue)
            self.filter.start()

        self.ingestor = Ingestor(self.ingestor_config, self.filter_input_queue)
        self.ingestor.start()
        self.log.info(log_msg.format("Started", self.app_name))

    def stop(self):
        """ Stops Video Ingestion pipeline
        """
        log_msg = "======={} {}======="
        self.log.info(log_msg.format("Stopping", self.app_name))
        os._exit(1)

    def _on_change_config_callback(self, key, value):
        """Callback method to be called by etcd

        :param key: Etcd key
        :type key: str
        :param value: Etcd value
        :type value: str
        """
        self.log.info("key: {}, value: {}".format(key, value))
        value = json.loads(value)
        ingestor_cfg = value["ingestor"]
        filter_cfg = value.get("filter", None)
        if filter_cfg != self.filter_config:
            self.log.info("Filter cfg: {}".format(filter_cfg))
            os._exit(0)
        if ingestor_cfg != self.ingestor_config:
            self.log.info("Ingestor cfg: {}".format(ingestor_cfg))
            self.ingestor.stop()
            self.ingestor_config = ingestor_cfg
            self.ingestor = Ingestor(self.ingestor_config,
                                     self.filter_input_queue)
            self.ingestor.start()
        else:
            self.log.info("Received same ingestor config...")


def main():
    """Main method
    """
    dev_mode = bool(strtobool(os.environ["DEV_MODE"]))
    conf = {
        "certFile": "",
        "keyFile": "",
        "trustFile": ""
    }
    if not dev_mode:
        conf = {
            "certFile": "/run/secrets/etcd_VideoIngestion_cert",
            "keyFile": "/run/secrets/etcd_VideoIngestion_key",
            "trustFile": "/run/secrets/ca_etcd"
        }
    cfg_mgr = ConfigManager()
    config_client = cfg_mgr.get_config_client("etcd", conf)

    log = configure_logging(os.environ['PY_LOG_LEVEL'].upper(),
                            __name__)

    vi = VideoIngestion(dev_mode, config_client)

    def handle_signal(signum, frame):
        log.info('Video Ingestion program killed...')
        vi.stop()

    signal.signal(signal.SIGTERM, handle_signal)

    try:
        vi.start()
    except Exception as ex:
        log.exception('Exception: {}'.format(ex))
        vi.stop()


if __name__ == '__main__':
    main()

File: src/video_ingestion.cpp
// Copyright (c) 2019 Intel Corporation.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM,OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.

/**
 * @file
 * @brief VideoIngestion Implementation
 */

#include <iostream>
#include <safe_lib.h>
#include "eii/vi/video_ingestion.h"
#include "eii/vi/ingestor.h"
#include "eii/vi/gstreamer_ingestor.h"
#include <mutex>

#define INTEL_VENDOR "GenuineIntel"
#define INTEL_VENDOR_LENGTH 12
#define DEFAULT_QUEUE_SIZE 10
#define PUB "pub"
#define SW_TRIGGER "sw_trigger"
#define ARGUMENTS "arguments"

using namespace eii::vi;
using namespace eii::utils;
using namespace eii::msgbus;
using namespace eii::udf;

VideoIngestion::VideoIngestion(
        std::string app_name, std::condition_variable& err_cv, char* vi_config, ConfigMgr* ctx, CommandHandler* commandhandler) :
    m_app_name(app_name), m_commandhandler(commandhandler), m_err_cv(err_cv), m_enc_type(EncodeType::NONE), m_enc_lvl(0) {

    // Parse the configuration
    config_t* config = json_config_new_from_buffer(vi_config);
    if (config == NULL) {
        const char* err = "Failed to initialize configuration object";
        LOG_ERROR("%s", err);
        throw(err);
    }
    config_value_t* encoding_value = config->get_config_value(config->cfg,
                                                              "encoding");
    if (encoding_value == NULL) {
        const char* err = "\"encoding\" key is missing";
        LOG_WARN("%s", err);
    } else {
        config_value_t* encoding_type_cvt = config_value_object_get(encoding_value,
                                                                    "type");
        if ( encoding_type_cvt == NULL ) {
            const char* err = "encoding \"type\" key missing";
            LOG_ERROR("%s", err);
            config_destroy(config);
            throw(err);
        }
        if (encoding_type_cvt->type != CVT_STRING) {
            const char* err = "encoding \"type\" value has to be of string type";
            LOG_ERROR("%s", err);
            config_destroy(config);
            config_value_destroy(encoding_type_cvt);
            throw(err);
        }
        char* enc_type = encoding_type_cvt->body.string;
        if (strcmp(enc_type, "jpeg") == 0) {
            m_enc_type = EncodeType::JPEG;
            LOG_DEBUG_0("Encoding type is jpeg");
        } else if (strcmp(enc_type, "png") == 0) {
            m_enc_type = EncodeType::PNG;
            LOG_DEBUG_0("Encoding type is png");
        } else {
            throw "Encoding type is not supported";
        }

        config_value_t* encoding_level_cvt = config_value_object_get(encoding_value,
                                                                    "level");
        if (encoding_level_cvt == NULL) {
            const char* err = "encoding \"level\" key missing";
            LOG_ERROR("%s", err);
            config_destroy(config);
            throw(err);
        }
        if (encoding_level_cvt->type != CVT_INTEGER) {
            const char* err = "encoding \"level\" value has to be of string type";
            LOG_ERROR("%s", err);
            config_destroy(config);
            config_value_destroy(encoding_level_cvt);
            throw(err);
        }
        m_enc_lvl = encoding_level_cvt->body.integer;
        LOG_DEBUG("Encoding value is %d", m_enc_lvl);
    }

    config_value_t* ingestor_value = config->get_config_value(config->cfg,
                                                              "ingestor");
    if (ingestor_value == NULL) {
        const char* err = "\"ingestor\" key is missing";
        LOG_ERROR("%s", err);
        throw(err);
    }
    config_value_t* ingestor_type_cvt = config_value_object_get(ingestor_value,
                                                                "type");
    if (ingestor_type_cvt == NULL) {
        const char* err = "\"type\" key missing";
        LOG_ERROR("%s", err);
        config_destroy(config);
        throw(err);
    }
    if (ingestor_type_cvt->type != CVT_STRING) {
        const char* err = "\"type\" value has to be of string type";
        LOG_ERROR("%s", err);
        config_destroy(config);
        config_value_destroy(ingestor_type_cvt);
        throw(err);
    }

    m_ingestor_type = std::string(ingestor_type_cvt->body.string);

    config_value_t* ingestor_queue_cvt = config_value_object_get(ingestor_value,
                                                                    "queue_size");
    size_t queue_size = DEFAULT_QUEUE_SIZE;
    if (ingestor_queue_cvt == NULL) {
        LOG_INFO("\"queue_size\" key missing, so using default queue size: \
                    %ld", queue_size);
    } else {
        if (ingestor_queue_cvt->type != CVT_INTEGER) {
            const char* err = "\"queue_size\" value has to be of integer type";
            LOG_ERROR("%s", err);
            config_destroy(config);
            config_value_destroy(ingestor_queue_cvt);
            throw(err);
        }
        queue_size = ingestor_queue_cvt->body.integer;
    }

    m_udf_input_queue = new FrameQueue(queue_size);

    config_value_object_t* ingestor_cvt = ingestor_value->body.object;
    m_ingestor_cfg = config_new(ingestor_cvt->object, free, get_config_value, NULL);
    if (m_ingestor_cfg == NULL) {
        const char* err = "Unable to get ingestor config";
        LOG_ERROR("%s", err);
        config_destroy(config);
        throw(err);
    }

    // get config SW_Trigger logic start
    config_value_t* sw_trigger = config->get_config_value(config->cfg,
                                                            SW_TRIGGER);
    if (sw_trigger == NULL) {
        LOG_INFO("Software Trigger feature is disabled");
        m_init_state_start = true;
        m_sw_trgr_en = false;
    } else {
        LOG_INFO("Software Trigger feature is enabled");
        m_sw_trgr_en = true;

        if (m_commandhandler != NULL) {
            m_commandhandler->register_callback((int)START_INGESTION, std::bind(&VideoIngestion::process_start_ingestion, this, std::placeholders::_1));
            m_commandhandler->register_callback((int)STOP_INGESTION, std::bind(&VideoIngestion::process_stop_ingestion, this, std::placeholders::_1));
            m_commandhandler->register_callback((int)SNAPSHOT, std::bind(&VideoIngestion::process_snapshot, this, std::placeholders::_1));
        }

        // Read config from config_mgr
        config_value_t* sw_trigger_init_state_cvt = config_value_object_get(sw_trigger,
                                                                "init_state");
        if (sw_trigger_init_state_cvt == NULL) {
            const char* err = "\"init_state\" key missing";
            LOG_ERROR("%s", err);
            config_destroy(config);
            throw(err);
        }

        // SW trigger init state : Possible values are "running"/"stopped"
        std::string initial_state = sw_trigger_init_state_cvt->body.string;

        // bool value of the init_state read from config
        if (initial_state == "running") {
            m_init_state_start = true;
        } else if (initial_state == "stopped") {
            m_init_state_start = false;
        } else {
            const char* err = "init_state value not supported";
            LOG_ERROR("\"%s\" %s", initial_state.c_str(), err);
            throw(err);
        }

        // Ingestion is not going ON by default when VideoIngestion constructor is called. Even if init_state=start, the ingestion
        // will start when start() is called
        m_ingestion_running.store(false);
    }

    config_value_t* udf_value = config->get_config_value(config->cfg,
                                                            "udfs");
    if (udf_value == NULL) {
        LOG_INFO("\"udfs\" key doesn't exist, so udf output queue is same as \
                udf input queue!!")
        m_udf_output_queue = m_udf_input_queue;
        m_udf_manager = NULL;
    } else {
        m_udf_output_queue = new FrameQueue(queue_size);
        m_udf_manager = new UdfManager(config, m_udf_input_queue, m_udf_output_queue, m_app_name,
                                        m_enc_type, m_enc_lvl);
    }

    // Get ingestor
    m_ingestor = get_ingestor( m_ingestor_cfg, m_udf_input_queue, m_ingestor_type.c_str(), m_app_name, m_snapshot_cv, m_enc_type, m_enc_lvl);

    PublisherCfg* pub_ctx = ctx->getPublisherByIndex(0);
    if (pub_ctx == NULL) {
        const char* err = "pub_ctx initialization failed";
        LOG_ERROR("%s", err);
        throw(err);
    }
    config_t* pub_config = pub_ctx->getMsgBusConfig();
    if (pub_config == NULL) {
        const char* err = "Failed to fetch msgbus config for Publisher";
        LOG_ERROR("%s", err);
        throw(err);
    }
    std::vector<std::string> topics = pub_ctx->getTopics();
    if (topics.empty()) {
        const char* err = "Topics list cannot be empty";
        LOG_ERROR("%s", err);
        throw(err);
    }
    LOG_DEBUG_0("Publisher Config received...");

    m_publisher = new PublisherThread(
            pub_config, m_err_cv, topics[0], (MessageQueue*) m_udf_output_queue, m_app_name);

    config_destroy(config);
    config_value_destroy(ingestor_type_cvt);
    config_value_destroy(ingestor_queue_cvt);
    config_value_destroy(encoding_value);
    config_value_destroy(udf_value);
}

msg_envelope_elem_body_t* VideoIngestion::process_start_ingestion(msg_envelope_elem_body_t *arg_payload) {
    try {
            LOG_INFO_0("START INGESTION request received from client");
            if (m_ingestion_running.load()) {
                std::string err = "Ingestion already running";
                return m_commandhandler->form_reply_payload((int)REQ_ALREADY_RUNNING, err, NULL);
            }
            IngestRetCode ret = m_ingestor->start();
            if (ret != IngestRetCode::SUCCESS) {
                LOG_ERROR("Failed to start ingestor thread: %d",ret);
                std::string err = "Failed to start ingestor thread";
                return m_commandhandler->form_reply_payload((int)REQ_NOT_HONORED, err, NULL);
            } else {
                LOG_INFO("Ingestor thread started...");
                m_ingestion_running.store(true);
                // acknowledging back to client that ingestion has actually started
                return m_commandhandler->form_reply_payload((int)REQ_HONORED, "SUCCESS", NULL);
            }
    } catch(std::exception& ex) {
        std::string err = "exception occurred request not honored";
        LOG_ERROR("%s %s", ex.what(), err.c_str());
        return m_commandhandler->form_reply_payload((int)REQ_NOT_HONORED, err, NULL);
    }
}

msg_envelope_elem_body_t* VideoIngestion::process_stop_ingestion(msg_envelope_elem_body_t *arg_payload) {
    try {
            LOG_INFO_0("STOP INGESTION request received from client");

            if (!m_ingestion_running.load()) {
                // form a JSON reply buffer
                std::string err = "Ingestion already stopped";
                return m_commandhandler->form_reply_payload((int)REQ_ALREADY_STOPPED, err, NULL);
            }
            // stop the ingestor
            if (m_ingestor) {
                m_ingestor->stop();
            }

            m_ingestion_running.store(false);

            // form a JSON reply buffer
            return m_commandhandler->form_reply_payload((int)REQ_HONORED, "SUCCESS", NULL);
    } catch(std::exception& ex) {
        std::string err = "exception occurred request not honored";
        LOG_ERROR("%s %s", ex.what(), err.c_str());
        return m_commandhandler->form_reply_payload((int)REQ_NOT_HONORED, err, NULL);
    }
}

msg_envelope_elem_body_t* VideoIngestion::process_snapshot(msg_envelope_elem_body_t *arg_payload) {
    try {
            LOG_INFO_0("SNAPSHOT request received from client");

            if (m_ingestion_running.load()) {
                std::string err = "Ingestion already running";
                return m_commandhandler->form_reply_payload((int)REQ_ALREADY_RUNNING, err, NULL);
            }

            IngestRetCode ret = IngestRetCode::NOT_INITIALIZED;
            if (m_ingestor) {
                ret = m_ingestor->start(true);
            }
            if (ret != IngestRetCode::SUCCESS) {
                LOG_ERROR("Failed to start ingestor thread: %d",ret);
                std::string err = "Failed to start ingestor thread";
                return m_commandhandler->form_reply_payload((int)REQ_NOT_HONORED, err, NULL);
            }
            LOG_DEBUG_0("Ingestor thread snapshot started");
            std::mutex mtx;
            std::unique_lock<std::mutex> lck(mtx);
            m_snapshot_cv.wait(lck);
            lck.unlock();
            LOG_DEBUG_0("Stopping ingestor thread");
            if (m_ingestor) {
                m_ingestor->stop();
            }
            m_ingestion_running.store(false);
            return m_commandhandler->form_reply_payload((int)REQ_HONORED, "SUCCESS", NULL);
    } catch(std::exception& ex) {
        std::string err = "exception occurred request not honored";
        LOG_ERROR("%s %s", ex.what(), err.c_str());
        return m_commandhandler->form_reply_payload((int)REQ_NOT_HONORED, err, NULL);
    }
}

VideoIngestion& VideoIngestion::operator=(const VideoIngestion& src) {
    return *this;
}

void VideoIngestion::start() {
    if (m_publisher) {
        m_publisher->start();
        LOG_INFO("Publisher thread started...");
    }
    if (m_udf_manager) {
        m_udf_manager->start();
        LOG_INFO("Started udf manager");
    }

    // if SW trigger is disabled OR (if sw trigger is enabled && init_state = running) then start ingestion
    if (!m_sw_trgr_en || (m_sw_trgr_en && m_init_state_start)) {
        IngestRetCode ret = m_ingestor->start();
        if (ret != IngestRetCode::SUCCESS) {
            LOG_ERROR_0("Failed to start ingestor thread");
        } else {
            LOG_INFO("Ingestor thread started...");
            m_ingestion_running.store((m_sw_trgr_en) ? true : false);
        }
    }
}

void VideoIngestion::stop() {
    if (m_ingestor) {
        m_ingestor->stop();
    }
    if (m_udf_manager) {
        m_udf_manager->stop();
    }
    if (m_publisher) {
        m_publisher->stop();
    }
}

VideoIngestion::~VideoIngestion() {
    // Stop the thread (if it is running)
    if (m_ingestor) {
        m_ingestor->stop();
    }
    if (m_ingestor) {
        delete m_ingestor;
    }
    if (m_udf_manager) {
        delete m_udf_manager;
    }
    if (m_publisher) {
        delete m_publisher;
    }
    if (m_udf_input_queue) {
        delete m_udf_input_queue;
    }
    if (m_udf_output_queue) {
        delete m_udf_output_queue;
    }
}
